### PR TITLE
Add first pass at GuestArray

### PR DESCRIPTION
**Pull request by kubkon**
While this is a WIP, I thought it might be useful to initiate a draft PR so that you can monitor the progress and shout out if you find anything amiss, wrong, etc.

### CODE REVIEW

File: crates/memory/src/memory.rs [hunk 1]
```diff
@@ -359,3 +359,72 @@ impl<'a, T> Drop for GuestRefMut<'a, T> {
...
+    pub fn as_ref(&'a self) -> Result<&'a [T], GuestError> {
+        // First, validate
+        for i in 0..self.num_elems {
+            let ptr = self.ptr.elem(i as i32)?;
+            T::validate(&ptr)?;
+        }
+        // Then, extract as slice
+        Ok(unsafe {
+            std::slice::from_raw_parts(self.ptr.as_raw() as *const T, self.num_elems as usize)
+        })
```
Comment (kubkon) [inline crates/memory/src/memory.rs]: I think this is slightly better now compared to what it was in the previous commit d910c0e. I hope I'm right in my thinking here that by firstly validating that all (expected) pointers are valid, we can then safely extract the slice `&'a [T]` utilising `std::slice::from_raw_parts` with a raw pointer cast.

Comment (pchickey) [reply crates/memory/src/memory.rs]: I believe this is correct as well.

File: crates/memory/src/memory.rs [hunk 2]
```diff
@@ -359,3 +359,267 @@ impl<'a, T> Drop for GuestRefMut<'a, T> {
...
+            num_elems: self.num_elems,
+        })
+    }
+}
+
+pub struct GuestArrayRef<'a, T>
+where
+    T: GuestType,
+{
+    r#ref: GuestRef<'a, T>,
```
[the remark this answers is not in the document]
Comment (kubkon) [reply crates/memory/src/memory.rs]: SGTM! I'll make sure to follow this style then from now on.

File: crates/memory/src/memory.rs [hunk 3]
```diff
@@ -359,3 +359,267 @@ impl<'a, T> Drop for GuestRefMut<'a, T> {
...
     }
 }
+
+pub struct GuestArray<'a, T>
+where
+    T: GuestType,
+{
+    ptr: GuestPtr<'a, T>,
+    num_elems: u32,
+}
```
[the remark this answers is not in the document]
Comment (kubkon) [reply crates/memory/src/memory.rs]: Excellent point, thanks! I think I'm also partial towards the latter. Let me see what I can conjure up!

File: crates/memory/src/memory.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kubkon) [reply crates/memory/src/memory.rs]: Done in 360aa71.

File: crates/memory/src/memory.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kubkon) [reply crates/memory/src/memory.rs]: Done in cc4f33f.

File: crates/memory/src/memory.rs [hunk 4]
```diff
@@ -359,3 +359,267 @@ impl<'a, T> Drop for GuestRefMut<'a, T> {
...
+    }
+}
+
+pub struct GuestArrayMut<'a, T>
+where
+    T: GuestType,
+{
+    ptr: GuestPtrMut<'a, T>,
+    num_elems: u32,
+}
```
[the remark this answers is not in the document]
Comment (kubkon) [reply crates/memory/src/memory.rs]: Done in cc4f33f.

**Comment by kubkon** [comment]
While addressing the comments (thanks!), I've gone ahead and implemented `fmt::Debug` for memory types. Note thought that the implementation is *not* enforced on the generic parameter `T` in the struct; rather, if `T` impls `fmt::Debug` then so does the memory type such as `GuestPtr<'_, T>`. This should simplify testing and debugging in general (I hope!).

**Pull request merged**